README.md
![Francium](Assets/logo.png)

Francium is part of the **[E-sites iOS Suite](https://github.com/e-sites/iOS-Suite)**.

---

A small library to use for your file system.

[![forthebadge](http://forthebadge.com/images/badges/made-with-swift.svg)](http://forthebadge.com) [![forthebadge](http://forthebadge.com/images/badges/built-with-swag.svg)](http://forthebadge.com)

[![Platform](https://img.shields.io/cocoapods/p/Francium.svg?style=flat)](http://cocoadocs.org/docsets/Francium)
[![CocoaPods Compatible](https://img.shields.io/cocoapods/v/Francium.svg)](http://cocoadocs.org/docsets/Francium)
[![Quality](https://apps.e-sites.nl/cocoapodsquality/Francium/badge.svg?004)](https://cocoapods.org/pods/Francium/quality)
[![Travis-ci](https://travis-ci.org/e-sites/Francium.svg?branch=master&001)](https://travis-ci.org/e-sites/Francium)


# Installation

Podfile:

```ruby
pod 'Francium'
```

And then

```
pod install
```

# Implementation

Both `File` and `Dir` implement the following functions / properties:

```swift

 /// The absolute path:
/// `File("path: ./").absolutePath #=> /var/folders/6d/FranciumTests-61688042-63CA-4E70-A9A6-31BBBB12FD38/`
public var absolutePath: String 

/// Does the entry exist?
public var isExisting: Bool

/// The base name
/// `File("path: /usr/bin/testfile.swift") #=> testfile.swift`
public var basename: String 

/// The file (or directory) name without the final extension
/// `File("path: testfile.rb").name #=> testfile`
public var name: String

public var isDirectory: Bool 

/// The name of the directory
/// `File("path: testfile.rb").dirname #=> /var/folders/6d`
/// `Dir("path: ./").dirname #=> /var/folders/6d`
public var dirName: String

public var modificationDate: Date?
public var creationDate: Date?

// ---- Path

public func delete() throws
public func rename(to newName: String) throws
public func move(to dir: Dir, newName: String? = nil) throws 
public func copy(to dir: Dir, newName: String? = nil) throws

// ---- Permissions

public func chmod(_ value: Int)

public var permissions: Int?
```

## `File`

```swift
/// The extension name
/// `File("path: testfile.rb").extensionName #=> rb`
public var extensionName: String

public static func create(path: String, permissions: Int = 0o0777) throws -> File
public func create() throws

public func write(data: Data) throws
public func write(string: String) throws
public func append(string: String) throws
public func append(data: Data) throws

public func read() throws -> Data
public var contents: String?

```

## `Dir`

```swift
public static func create(path: String,
                          withIntermediateDirectories: Bool = true,
                          permissions: Int = 0o0777) throws -> Dir

public func make(withIntermediateDirectories: Bool = true) throws

public func glob(_ pattern: String) -> [File]
public func empty(recursively: Bool = true) throws
```

CHANGELOG.md
# Changelog Francium

## v1.1.1 (18-10-2021)
- Swift 5.5

## v1.1.0 (05-04-2019)
- Swift 5.0

## v1.0.1 (13-07-2018)
- SPM support

## v1.0 (13-07-2018)
- Initial release

Tests/FranciumTests/FranciumTests.swift
//
//  FranciumTests.swift
//  FranciumTests
//
//  Created by Bas van Kuijck on 13/07/2018.
//  Copyright © 2018 E-sites. All rights reserved.
//

import XCTest
@testable import Francium

class FranciumTests: XCTestCase {
    
    override func setUp() {
        super.setUp()
    }

    private func _createTestDirectory() {

    }

    func testFile() {
        let file = File(path: "somenewfile.txt")
        XCTAssertEqual(file.isExisting, false)
        XCTAssertEqual(file.basename, "somenewfile.txt")
        XCTAssertEqual(file.name, "somenewfile")
        XCTAssertEqual(file.extensionName, "txt")
        XCTAssertNil(file.creationDate)
        XCTAssertNil(file.modificationDate)
    }

    func testDir() {
        let dir = Dir(path: "./")
        XCTAssertEqual(dir.isExisting, true)
        XCTAssertEqual(dir.isDirectory, true)
        XCTAssertNotNil(dir.creationDate)
        XCTAssertNotNil(dir.modificationDate)
    }

    func testCreation() {
        do {
            let dir = try Dir.create(path: "./SomeNewDirectory")
            XCTAssertNotNil(dir.creationDate)
            XCTAssertEqual(dir.permissions, 0o0777)
            XCTAssertEqual(dir.basename, "SomeNewDirectory")

            let file = try File.create(path: "./SomeNewDirectory/somenewfile.txt")
            XCTAssertEqual(file.permissions, 0o0777)
            XCTAssertNotNil(file.creationDate)
            XCTAssertEqual(file.basename, "somenewfile.txt")
            try file.write(string: "Some information")
            XCTAssertEqual(file.contents, "Some information")

            XCTAssertEqual(dir.glob("*").count, 1)
            XCTAssertEqual(dir.glob("some*").count, 1)
            XCTAssertEqual(dir.glob("*.txt").count, 1)
            XCTAssertEqual(dir.glob("*.md").count, 0)
            try file.delete()
            try dir.delete()
        } catch let error {
            XCTAssert(false, "\(error)")
        }
    }

    func testEmpty() {
        do {
            let dir = try Dir.create(path: "./SomeNewDirectory")
            let file = try File.create(path: "./SomeNewDirectory/somenewfile.txt")
            try dir.empty()
            XCTAssertEqual(dir.glob("*").count, 0)
            XCTAssertEqual(file.isExisting, false)
            try dir.delete()

        } catch let error {
            XCTAssert(false, "\(error)")
        }
    }

    func testAppend() {
        do {
            let file = try File.create(path: "./somenewfile.txt")
            try file.write(string: "Some information")
            XCTAssertEqual(file.contents, "Some information")
            try file.append(string: ">>>")
            XCTAssertEqual(file.contents, "Some information>>>")

            try file.delete()

        } catch let error {
            XCTAssert(false, "\(error)")
        }
    }
}


Francium/IO/File.swift
//
//  File.swift
//  Francium
//
//  Created by Bas van Kuijck on 13/07/2018.
//  Copyright © 2018 E-sites. All rights reserved.
//

import Foundation

public class File: IOObject, CustomStringConvertible {
    /// The extension name
    /// `File("path: testfile.rb").extensionName #=> rb`
    public var extensionName: String {
        let expl = url.lastPathComponent.components(separatedBy: ".")
        return expl.last ?? ""
    }

    public static func create(path: String, permissions: Int = 0o0777) throws -> File {
        let file = File(path: path)
        try file.create()
        file.chmod(permissions)
        return file
    }

    public func create() throws {
        if isExisting {
            throw Error.alreadyExists
        }

        let attributes: [FileAttributeKey: Any] = [
            .posixPermissions: 0o777
        ]

        if !FileManager.default.createFile(atPath: absolutePath, contents: nil, attributes: attributes) {
            throw Error.cannotCreateFile
        }
    }

    public var description: String {
        return "<File> [ absolutePath: \(absolutePath) ]"
    }
}

extension File {
    public func write(data: Data) throws {
        try data.write(to: url)
    }

    public func write(string: String) throws {
        let data = string.data(using: .utf8) ?? Data()
        try write(data: data)
    }

    public func append(string: String) throws {
        let data = string.data(using: .utf8) ?? Data()
        try append(data: data)
    }

    public func append(data: Data) throws {
        if !isExisting {
            try write(data: data)
            return
        }

        guard let fileHandle = FileHandle(forUpdatingAtPath: path) else {
            try write(data: data)
            return
        }

        defer {
            fileHandle.closeFile()
        }
        fileHandle.seekToEndOfFile()
        fileHandle.write(data)
    }

    public func read() throws -> Data {
        return try Data(contentsOf: url)
    }

    public var contents: String? {
        do {
            let data = try read()
            return String(data: data, encoding: .utf8)
        } catch {
            return nil
        }
    }
}


Sources/Francium/IO/Dir.swift
//
//  Dir.swift
//  Francium
//
//  Created by Bas van Kuijck on 13/07/2018.
//  Copyright © 2018 E-sites. All rights reserved.
//

#if os(Linux)
import Glibc

let system_glob = Glibc.glob
#else
import Darwin

let system_glob = Darwin.glob
#endif

import Foundation

public class Dir: IOObject, CustomStringConvertible {
    
    override public var isExisting: Bool {
        return isDirectory && super.isExisting
    }

    public static func create(path: String,
                              withIntermediateDirectories: Bool = true,
                              permissions: Int = 0o0777) throws -> Dir {
        let dir = Dir(path: path)
        try dir.make(withIntermediateDirectories: withIntermediateDirectories)
        dir.chmod(permissions)
        return dir
    }

    public func make(withIntermediateDirectories: Bool = true) throws {
        if isDirectory {
            throw Error.alreadyExists
        }

        let attributes: [FileAttributeKey: Any] = [
            .posixPermissions: 0o777
        ]
        try FileManager.default.createDirectory(atPath: absolutePath,
                                                withIntermediateDirectories: withIntermediateDirectories,
                                                attributes: attributes)
    }

    public var description: String {
        return "<Dir> [ absolutePath: \(absolutePath) ]"
    }

    public func glob(_ pattern: String) -> [File] {
        let pattern = absolutePath + "/" + pattern
        guard let cPattern = strdup(pattern) else {
            return []
        }
        var gt = glob_t()
        defer {
            globfree(&gt)
            free(cPattern)
        }

        let flags = GLOB_TILDE | GLOB_BRACE | GLOB_MARK
        if system_glob(cPattern, flags, nil, &gt) == 0 {
            #if os(Linux)
            let matchc = gt.gl_pathc
            #else
            let matchc = gt.gl_matchc
            #endif

            return (0..<Int(matchc)).compactMap { index in
                if let path = String(validatingUTF8: gt.gl_pathv[index]!) {
                    return File(path: path)
                }

                return nil
            }
        }
        return []
    }

    public func empty(recursively: Bool = true) throws {
        let fileManager = FileManager.default
        let files = try fileManager.contentsOfDirectory(atPath: absolutePath)
            .map { File(path: "\(absolutePath)/\($0)") }
        for file in files {
            if file.isDirectory {
                let dir = Dir(path: file.path)
                if recursively {
                    try dir.empty(recursively: true)
                } else {
                    if !dir.glob("*").isEmpty {
                        throw Error.directoryNotEmpty
                    }
                }
            }
            try file.delete()
        }
    }
}


Sources/Francium/IO/IOObject.swift
//
//  Pointer.swift
//  Francium
//
//  Created by Bas van Kuijck on 13/07/2018.
//  Copyright © 2018 E-sites. All rights reserved.
//

import Foundation

public class IOObject {

    public let path: String
    let url: URL

    public init(path: String) {
        self.path = path
        self.url = URL(fileURLWithPath: path)
    }

    /// The absolute path:
    /// `File("path: ./").absolutePath #=> /var/folders/6d/FranciumTests-61688042-63CA-4E70-A9A6-31BBBB12FD38/`
    public var absolutePath: String {
        return url.path
    }

    /// Does the entry exist?
    public var isExisting: Bool {
        return FileManager.default.fileExists(atPath: absolutePath)
    }

    /// The base name
    /// `File("path: /usr/bin/testfile.swift") #=> testfile.swift`
    public var basename: String {
        return url.lastPathComponent
    }

    /// The file (or directory) name without the final extension
    /// `File("path: testfile.rb").name #=> testfile`
    public var name: String {
        var expl = url.lastPathComponent.components(separatedBy: ".")
        if expl.count > 1 {
            expl.removeLast()
        }
        return expl.joined(separator: ".")
    }

    public var isDirectory: Bool {
        var isDir: ObjCBool = false
        if FileManager.default.fileExists(atPath: path, isDirectory: &isDir) {
            return isDir.boolValue
        } else {
            return false
        }
    }

    /// The name of the directory
    /// `File("path: testfile.rb").dirname #=> /var/folders/6d`
    /// `Dir("path: ./").dirname #=> /var/folders/6d`
    public var dirName: String {
        var components = url.pathComponents
        if !isDirectory {
            components.removeLast()
        }

        if components.last == "" {
            components.removeLast()
        }
        return components.joined(separator: "/")
    }


    public var modificationDate: Date? {
        return _attribute(type: Date.self, for: .modificationDate)
    }

    public var creationDate: Date? {
        return _attribute(type: Date.self, for: .creationDate)
    }
}

extension IOObject {
    public func delete() throws {
        try FileManager.default.removeItem(atPath: absolutePath)
    }
}

extension IOObject {
    public func rename(to newName: String) throws {
        try move(to: Dir(path: dirName), newName: newName)
    }

    public func move(to dir: Dir, newName: String? = nil) throws {
        try FileManager.default.moveItem(atPath: absolutePath, toPath: dir.dirName + "/" + (newName ?? basename))
    }

    public func copy(to dir: Dir, newName: String? = nil) throws {
        try FileManager.default.copyItem(atPath: absolutePath, toPath: dir.dirName + "/" + (newName ?? basename))
    }
}

extension IOObject {
    public func chmod(_ value: Int) {
        _setAttribute(value: value, for: .posixPermissions)
    }

    public var permissions: Int? {
        return _attribute(type: Int.self, for: .posixPermissions)
    }
}

extension IOObject {
    fileprivate func _attribute<T>(type: T.Type, `for` key: FileAttributeKey) -> T? {
        do {
            let attributes = try FileManager.default.attributesOfItem(atPath: absolutePath)
            return attributes[key] as? T

        } catch {
            return nil
        }
    }

    fileprivate func _setAttribute(value: Any, for key: FileAttributeKey) {
        do {
            var attributes = try FileManager.default.attributesOfItem(atPath: absolutePath)
            attributes[key] = value
            try FileManager.default.setAttributes(attributes, ofItemAtPath: absolutePath)
        } catch {
        }
    }
}

